design-proposals/instancetype.kubevirt.io/pkg-instancetype-refactor.md

```diff
@@ -0,0 +1,160 @@
+# Overview
+
+The implementation of instance types (and formally flavors) started with KubeVirt v0.46.0 and has slowly grown in scope and features. The structure of the code however hasn’t changed much since the initial implementation of flavors and as a result is mostly concentrated within a handful of files within `pkg/instancetype`. The initial design decision of providing a single public `interface` and implementation also has not scaled well as new features and functionality have been added to the codebase.
+
+## Motivation
+
+Ultimately the instance type code base is becoming more difficult to both maintain and extend. Refactoring the code to make it more manageable for future contributors and maintainers ahead of the API graduating to `v1` is extremely important as a result.
+
+## Goals
+
+* Break up `pkg/instancetype/instancetype.go` and `pkg/instancetype/instancetype_test.go`  
+* Provide finer grain implementations of common instance type and preference functionality such as find, apply, store and infer  
+* Replace the public `InstancetypesMethods` interface with private smaller interfaces defined closer to their consumers
+* All existing functionality should be unchanged by this work, all functional tests should therefore be untouched
+
+## Non Goals
+
+* Any behavioral changes to the API during the release v1.4 cycle should happen outside of this work
+
+## User Stories
+
+* As a developer working on `pkg/instancetype` I want to see the code logically separated for easier maintenance and future improvements  
+* As a developer working on `pkg/instancetype` I want to see the API graduate to `v1`
```

Comment:
It is not part of this proposal though?

Reply:
In my mind this refactor is something we need to do ahead of the API hitting v1 but I hadn't made that explicit anywhere else in the doc. I'll add something to goals etc.